{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "lab05.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "colab_type": "code",
        "id": "Ag9FshAu2pce",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 231
        },
        "outputId": "b4fd9c38-b0f5-4b1b-a839-4fc5116dbd84"
      },
      "source": [
        "# Transcibe\n",
        "def transcribe1(seq): \n",
        "  RNA_seq = seq.replace('T', 'U')\n",
        "  ### Define location of start codon (AUG)\n",
        "  start_codon = RNA_seq.find('AUG')\n",
        "  codons = []\n",
        "  for i in range(start_codon, len(RNA_seq), 3): \n",
        "    codon = RNA_seq[i:i + 3]\n",
        "    if len(codon)==3:\n",
        "        codons.append(codon)\n",
        "  return codons\n",
        "# Lab 4 Origin:  \n",
        "\n",
        "def translation(DNA_seq):\n",
        "  mRNA= transcribe(DNA_seq)\n",
        "  one_letter=''\n",
        "  for n in range (0, len(mRNA)):\n",
        "    if mRNA[n] =='UAA' or mRNA[n]=='UGA'or mRNA[n] =='UAG':\n",
        "      break\n",
        "    for key1, three_letter in translate.items():\n",
        "      if mRNA[n] == key1:\n",
        "        for key2, value2 in AA_mapping.items():\n",
        "          if three_letter == key2:\n",
        "            one_letter = one_letter + value2\n",
        "  return (one_letter)\n",
        "\n",
        "protein_seq= translation(given_seq)\n",
        "print (protein_seq)\n",
        "\n",
        "# Dr. Jie Hou's way:\n",
        "\n",
        "def translation(DNA_seq):\n",
        "  mRNA= transcribe(DNA_seq)\n",
        "  one_letter=''\n",
        "\n",
        "  for codon in mRNA:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      return one_letter\n",
        "    one_letter = one_letter +  AA_mapping[translate[codon]]\n",
        "  return (one_letter)\n",
        "\n",
        "protein_seq= translation(given_seq)\n",
        "print (protein_seq)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-74-096a748336a2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     14\u001b[0m   \u001b[0;32mreturn\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mone_letter\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0mprotein_seq\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mtranslation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgiven_seq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mprotein_seq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'given_seq' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "wLjcIx_7nXEv",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#######################################################################################\n",
        "# (1) define dictionary to translate three-letter mRNA to three-letter amino acid here\n",
        "#######################################################################################\n",
        "translate = { \n",
        "  'UUU':'Phe','UCU':'Ser','UAU':'Tyr','UGU':'Cys',\n",
        "  'UUC':'Phe','UCC':'Ser','UAC':'Tyr','UGC':'Cys',\n",
        "  'UUA':'Leu','UCA':'Ser','UAA':'Stop','UGA':'Stop',\n",
        "  'UUG':'Leu','UCG':'Ser','UAG':'Stop','UGG':'Trp',\n",
        "  'CUU':'Leu','CCU':'Pro','CAU':'His','CGU':'Arg',\n",
        "  'CUC':'Leu','CCC':'Pro','CAC':'His','CGC':'Arg',\n",
        "  'CUA':'Leu','CCA':'Pro','CAA':'Gln','CGA':'Arg',\n",
        "  'CUG':'Leu','CCG':'Pro','CAG':'Gln','CGG':'Arg',\n",
        "  'AUU':'Ile','ACU':'Thr','AAU':'Asn','AGU':'Ser',\n",
        "  'AUC':'Ile','ACC':'Thr','AAC':'Asn','AGC':'Ser',\n",
        "  'AUA':'Ile','ACA':'Thr','AAA':'Lys','AGA':'Arg',\n",
        "  'AUG':'Met','ACG':'Thr','AAG':'Lys','AGG':'Arg',\n",
        "  'GUU':'Val','GCU':'Ala','GAU':'Asp','GGU':'Gly',\n",
        "  'GUC':'Val','GCC':'Ala','GAC':'Asp','GGC':'Gly',\n",
        "  'GUA':'Val','GCA':'Ala','GAA':'Glu','GGA':'Gly',\n",
        "  'GUG':'Val','GCG':'Ala','GAG':'Glu','GGG':'Gly'\n",
        "}\n",
        "\n",
        "#######################################################################################\n",
        "# (2) define dictionary to map three-letter amino acid to one-letter amino acid here\n",
        "#######################################################################################\n",
        "AA_mapping = {\n",
        "    'Ala':'A', 'Arg':'R','Asn':'N',\n",
        "    'Asp':'D', 'Cys':'C','Gln':'Q',\n",
        "    'Glu':'E', 'Gly':'G','His':'H',\n",
        "    'Ile':'I', 'Leu':'L','Lys':'K',\n",
        "    'Met':'M', 'Phe':'F','Pro':'P',\n",
        "    'Ser':'S', 'Thr':'T','Trp':'W',\n",
        "    'Tyr':'Y','Val':'V'\n",
        "}\n",
        "#######################################################################################\n",
        "# (3) define dictionary to map base_pairing in RNA\n",
        "#######################################################################################\n",
        "base_pairing = { 'G':'C','C':'G','U':'A','A':'U'}"
      ],
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WZz0CK8OnSYT",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        " 'Only run 1 if going up to question 5' \n",
        "'''\n",
        "This is the function in which you must place your code to implement DNA/RNA translation pipeline.\n",
        "The input is any DNA sequence.\n",
        "The return type should be the protein sequence translated from DNA sequence.\n",
        "'''\n",
        "def transcribe(seq): \n",
        "  RNA_seq = seq.replace('T', 'U')\n",
        "  ### Define location of start codon (AUG)\n",
        "  start_codon = RNA_seq.find('AUG')\n",
        "  codons = []\n",
        "  for i in range(start_codon, len(RNA_seq), 3): \n",
        "    codon = RNA_seq[i:i + 3]\n",
        "    if len(codon)==3:\n",
        "        codons.append(codon)\n",
        "  return codons\n",
        "\n",
        "def get_protein_from_DNA(DNA):\n",
        "  protein = ''\n",
        "  mRNA= transcribe(DNA)\n",
        "  for n in range (0, len(mRNA)):\n",
        "    if mRNA[n] =='UAA' or mRNA[n]=='UGA'or mRNA[n] =='UAG':\n",
        "      break\n",
        "    for key1, three_letter in translate.items():\n",
        "      if mRNA[n] == key1:\n",
        "        for key2, value2 in AA_mapping.items():\n",
        "          if three_letter == key2:\n",
        "            protein +=  value2\n",
        "  return protein\n",
        "\n",
        "\n",
        "  ### start your implementation\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab_type": "code",
        "id": "MUIRmjer8l3m",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "outputId": "4c21fa82-7336-40d1-b7f0-35b95c8360df"
      },
      "source": [
        "# Scrap\n",
        "# Transcribe function testing\n",
        "# This code works !!!\n",
        "def transcribe(seq): \n",
        "  RNA_seq = seq.replace('T', 'U')\n",
        "  codons1 = []\n",
        "  codons2 = []\n",
        "  codons3 = []\n",
        "  codons4 = []\n",
        "  codons5 = []\n",
        "  codons6 = []\n",
        "  for i in range(0, len(RNA_seq), 3): \n",
        "    codon1 = RNA_seq[i:i + 3]\n",
        "    if len(codon1)==3:\n",
        "        codons1.append(codon1)\n",
        "  for i in range(1, len(RNA_seq), 3): \n",
        "    codon2 = RNA_seq[i:i + 3]\n",
        "    if len(codon2)==3:\n",
        "        codons2.append(codon2)\n",
        "  for i in range(2, len(RNA_seq), 3): \n",
        "    codon3 = RNA_seq[i:i + 3]\n",
        "    if len(codon3)==3:\n",
        "        codons3.append(codon3)\n",
        "  cDNAf1 = ''\n",
        "  for i in range(0, len(RNA_seq)):\n",
        "    cDNAf1 += base_pairing[RNA_seq[i]]\n",
        "  for i in range(0, len(cDNAf1), 3): \n",
        "    codon4 = cDNAf1[i:i + 3]\n",
        "    if len(codon4)==3:\n",
        "      codons4.append(codon4)\n",
        "  cDNAf2 = ''\n",
        "  for i in range(1, len(RNA_seq)):\n",
        "    cDNAf2 += base_pairing[RNA_seq[i]]\n",
        "  for i in range(0, len(cDNAf2), 3): \n",
        "    codon5 = cDNAf2[i:i + 3]\n",
        "    if len(codon5)==3:\n",
        "      codons5.append(codon5)    \n",
        "  cDNAf3 = ''\n",
        "  for i in range(2, len(RNA_seq)):\n",
        "    cDNAf3 += base_pairing[RNA_seq[i]]\n",
        "  for i in range(0, len(cDNAf3), 3): \n",
        "    codon6 = cDNAf3[i:i + 3]\n",
        "    if len(codon6)==3:\n",
        "      codons6.append(codon6) \n",
        "  return codons1 , codons2 , codons3, codons4, codons5, codons6\n",
        "\n",
        "\n",
        "#Complement DNA -> mRNA\n",
        "#UAC AGA UGA GUG CUA UGU AGG GAC UUU U\n",
        "#ACA GAU GAG UGC UAU GUA GGG ACU UUU\n",
        "#CAG AUG AGU GCU AUG UAG GGA CUU UU"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "['CAG', 'AUG', 'AGU', 'GCU', 'AUG', 'UAG', 'GGA', 'CUU']\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "64cc-I55Evqe",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 54
        },
        "outputId": "7bbbba7f-034e-4396-8e6c-c1de4901478e"
      },
      "source": [
        "# Scrap\n",
        "# Transcribe testing \n",
        "def transcribe(seq):\n",
        "  RNA_seq = seq.replace('T', 'U')\n",
        "  codons1 = []\n",
        "  codons4 = []\n",
        "  for i in range(0, len(RNA_seq), 3): \n",
        "    codon1 = RNA_seq[i:i + 3]\n",
        "    if len(codon1)==3:\n",
        "      codons1.append(codon1)\n",
        "\n",
        "\n",
        "  \n",
        "  for i in range(0, len(RNA_seq)-1, 3):\n",
        "    codon4=''\n",
        "    codon4 += base_pairing[RNA_seq[i]] + base_pairing[RNA_seq[i+1]] + base_pairing[RNA_seq[i+2]]\n",
        "    codon4 = sorted(''.join(codon4), reverse =True)\n",
        "    if len(codon4)==3:\n",
        "      codons4.append(codon4)\n",
        "  return codons1,codons4\n",
        "\n",
        "y=transcribe(test_seq)\n",
        "\n",
        "print (y)\n",
        "#UACAGAUGAGUGCUAUGUAGGGACUUUU"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "(['AUG', 'UCU', 'ACU', 'CAC', 'GAU', 'ACA', 'UCC', 'CUG', 'AAA'], [['U', 'C', 'A'], ['G', 'A', 'A'], ['U', 'G', 'A'], ['U', 'G', 'G'], ['U', 'C', 'A'], ['U', 'U', 'G'], ['G', 'G', 'A'], ['G', 'C', 'A'], ['U', 'U', 'U']])\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "05fjaLwnEjNN",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "# Scrap\n",
        "# Translate testing \n",
        "def get_protein_from_DNA(DNA):\n",
        "  mRNA=transcribe(DNA)\n",
        "  mRNA1= mRNA[0]\n",
        "  mRNA2= mRNA[1]\n",
        "  mRNA3= mRNA[2]\n",
        "  one_letter1=''\n",
        "  one_letter2=''\n",
        "  one_letter3=''\n",
        "  for codon in mRNA1:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue\n",
        "    one_letter1 +=  AA_mapping[translate[codon]]\n",
        "  for codon in mRNA2:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue\n",
        "    one_letter2 +=  AA_mapping[translate[codon]] \n",
        "  for codon in mRNA3:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue \n",
        "    one_letter3 +=  AA_mapping[translate[codon]]  \n",
        "  Result =  \" 5'to 3' Frame 1:\", one_letter1, \" \\n5'to 3' Frame 2:\", one_letter2, \" \\n5'to 3' Frame 3:\", one_letter3\n",
        "  return Result\n",
        "\n",
        "\n",
        "y = get_protein_from_DNA (test_seq) \n",
        "print (y)\n",
        "#Note: try to use \\n to create new line, cannot be used without print. Need to figure \\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab_type": "code",
        "id": "0glWYvZBAkre",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 221
        },
        "outputId": "b138194e-7500-4022-d473-db2ee3eb307d"
      },
      "source": [
        "' Only run 2 if going up to question 6 _ Testing ' \n",
        "'''\n",
        "This is the function in which you must place your code to implement DNA/RNA translation pipeline.\n",
        "The input is any DNA sequence.\n",
        "The return type should be the protein sequence translated from DNA sequence.\n",
        "'''\n",
        "\n",
        "#'start_codon = RNA_seq.find('AUG')'\n",
        "  ### Define location of start codon (AUG)\n",
        "\n",
        "def transcribe(seq): \n",
        "  RNA_seq = seq.replace('T', 'U')\n",
        "  codons1 = []\n",
        "  codons2 = []\n",
        "  codons3 = []\n",
        "  codons4 = []\n",
        "  codons5 = []\n",
        "  codons6 = []\n",
        "  for i in range(0, len(RNA_seq), 3): \n",
        "    codon1 = RNA_seq[i:i + 3]\n",
        "    if len(codon1)==3:\n",
        "        codons1.append(codon1)\n",
        "  for i in range(1, len(RNA_seq), 3): \n",
        "    codon2 = RNA_seq[i:i + 3]\n",
        "    if len(codon2)==3:\n",
        "        codons2.append(codon2)\n",
        "  for i in range(2, len(RNA_seq), 3): \n",
        "    codon3 = RNA_seq[i:i + 3]\n",
        "    if len(codon3)==3:\n",
        "        codons3.append(codon3)\n",
        "  cDNAf1 = ''\n",
        "  for i in range(0, len(RNA_seq)):\n",
        "    cDNAf1 += base_pairing[RNA_seq[i]]\n",
        "  for i in range(0, len(cDNAf1), 3): \n",
        "    codon4 = cDNAf1[i:i + 3]\n",
        "    if len(codon4)==3:\n",
        "      codons4.append(codon4)\n",
        "  cDNAf2 = ''\n",
        "  for i in range(1, len(RNA_seq)):\n",
        "    cDNAf2 += base_pairing[RNA_seq[i]]\n",
        "  for i in range(0, len(cDNAf2), 3): \n",
        "    codon5 = cDNAf2[i:i + 3]\n",
        "    if len(codon5)==3:\n",
        "      codons5.append(codon5)    \n",
        "  cDNAf3 = ''\n",
        "  for i in range(2, len(RNA_seq)):\n",
        "    cDNAf3 += base_pairing[RNA_seq[i]]\n",
        "  for i in range(0, len(cDNAf3), 3): \n",
        "    codon6 = cDNAf3[i:i + 3]\n",
        "    if len(codon6)==3:\n",
        "      codons6.append(codon6) \n",
        "  return codons1 , codons2 , codons3, codons4, codons5, codons6\n",
        "\n",
        "\n",
        "\n",
        "# I decide yo use Dr. Jie Hou to make it easier to code \n",
        "\n",
        "def get_protein_from_DNA(DNA):\n",
        "  mRNA=transcribe(DNA)\n",
        "  mRNA1= mRNA[0]\n",
        "  mRNA2= mRNA[1]\n",
        "  mRNA3= mRNA[2]\n",
        "  mRNA4= mRNA[3]\n",
        "  mRNA5= mRNA[4]\n",
        "  mRNA6= mRNA[5]\n",
        "  one_letter1=''\n",
        "  one_letter2=''\n",
        "  one_letter3=''\n",
        "  one_letter4=''\n",
        "  one_letter5=''\n",
        "  one_letter6=''\n",
        "  for codon in mRNA1:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue\n",
        "    one_letter1 +=  AA_mapping[translate[codon]]\n",
        "  for codon in mRNA2:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue\n",
        "    one_letter2 +=  AA_mapping[translate[codon]] \n",
        "  for codon in mRNA3:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue \n",
        "    one_letter3 +=  AA_mapping[translate[codon]] \n",
        "  for codon in mRNA4:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue \n",
        "    one_letter4 +=  AA_mapping[translate[codon]] \n",
        "  for codon in mRNA5:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue \n",
        "    one_letter5 +=  AA_mapping[translate[codon]] \n",
        "  for codon in mRNA6:\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      continue\n",
        "    one_letter6 +=  AA_mapping[translate[codon]] \n",
        "  if one_letter1.find('M') != -1:\n",
        "    one_letter1 = '*' + one_letter1 \n",
        "  \n",
        "  if one_letter2.find('M') != -1:\n",
        "    one_letter2 = '*' + one_letter2 \n",
        "  \n",
        "  if one_letter3.find('M') != -1:\n",
        "    one_letter3 = '*' + one_letter3 \n",
        "  \n",
        "  if one_letter4.find('M') != -1:\n",
        "    one_letter4 = '*' + one_letter4 \n",
        "  \n",
        "  if one_letter5.find('M') != -1:\n",
        "    one_letter5 = '*' + one_letter5 \n",
        "  \n",
        "  if one_letter6.find('M') != -1:\n",
        "    one_letter6 = '*' + one_letter6 \n",
        "  \n",
        "  r = \" * means ORF sequence; \\n5'to 3' Frame 1:\", one_letter1, \"\\n5'to 3' Frame 2:\", one_letter2, \"\\n5'to 3' Frame 3:\", one_letter3, \"\\n3'to 5' Frame 1:\", one_letter4, \"\\n3'to 5' Frame 2:\", one_letter5, \"\\n3'to 5' Frame 3\", one_letter6 \n",
        "  return r\n",
        "\n",
        "r = get_protein_from_DNA(test_seq)\n",
        "r\n",
        "  ### start your implementation\n",
        "#Note: try to use \\n to create new line, cannot be used without print. Need to figure \\n\n"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(\" * means ORF sequence; \\n5'to 3' Frame 1:\",\n",
              " '*MSTHDTSLK',\n",
              " \"\\n5'to 3' Frame 2:\",\n",
              " 'CLLTIHPK',\n",
              " \"\\n5'to 3' Frame 3:\",\n",
              " 'VYSRYIPE',\n",
              " \"\\n3'to 5' Frame 1:\",\n",
              " 'YRVLCRDF',\n",
              " \"\\n3'to 5' Frame 2:\",\n",
              " 'TDECYVGTF',\n",
              " \"\\n3'to 5' Frame 3\",\n",
              " '*QMSAMGL')"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "rLg2EUAZouLp",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 52
        },
        "outputId": "442bbbef-a00e-49ee-d7a2-7993b656a905"
      },
      "source": [
        "''' This is for testing only to make sure your function is implemented correctly. ''' \n",
        "test_seq = 'ATGTCTACTCACGATACATCCCTGAAAA' \n",
        "\n",
        "get_protein_from_DNA(test_seq) \n",
        "\n",
        "### expected result is MSTHDTSLK"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "\" * means ORF sequence; \\n5'to 3' Frame 1:*<MSTHDTSLK\\n5'to 3' Frame 2:CLLTIHPK\\n5'to 3' Frame 3:VYSRYIPE\\n3'to 5' Frame 1:*FSG<MYRE>T\\n3'to 5' Frame 2:FQGCIVSRH\\n3'to 5' Frame 3:FRDVSVD\""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "osXFC6u0k1q_",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "outputId": "6f677309-dc28-4535-cfe2-2beec95ca305"
      },
      "source": [
        "# Meant to be run from up to question 6\n",
        "# Dr. Jie Hou's improvement for the code!\n",
        "\n",
        "## This is a trick for you to print the string with a highlighted regions\n",
        "def highlight_text(string_in):\n",
        "  string_in=string_in.replace('<','\\033[1m\\033[94m')\n",
        "  string_in=string_in.replace('>','\\033[0m')\n",
        "  print(string_in)\n",
        "\n",
        "def transcribe(seq): \n",
        "  RNA_seq = seq.replace('T', 'U')\n",
        "  codons1 = []\n",
        "  codons2 = []\n",
        "  codons3 = []\n",
        "  codons4 = []\n",
        "  codons5 = []\n",
        "  codons6 = []\n",
        "  for i in range(0, len(RNA_seq), 3): \n",
        "    codon1 = RNA_seq[i:i + 3]\n",
        "    if len(codon1)==3:\n",
        "        codons1.append(codon1)\n",
        "  for i in range(1, len(RNA_seq), 3): \n",
        "    codon2 = RNA_seq[i:i + 3]\n",
        "    if len(codon2)==3:\n",
        "        codons2.append(codon2)\n",
        "  for i in range(2, len(RNA_seq), 3): \n",
        "    codon3 = RNA_seq[i:i + 3]\n",
        "    if len(codon3)==3:\n",
        "        codons3.append(codon3)\n",
        "  \n",
        "  RNA_seq_rev = RNA_seq[::-1]    # you need reverse the DNA sequence first, and then do the complement\n",
        "  cDNAf1 = ''\n",
        "  for i in range(0, len(RNA_seq_rev)):\n",
        "    cDNAf1 += base_pairing[RNA_seq_rev[i]]\n",
        "  for i in range(0, len(cDNAf1), 3): \n",
        "    codon4 = cDNAf1[i:i + 3]\n",
        "    if len(codon4)==3:\n",
        "      codons4.append(codon4)\n",
        "  cDNAf2 = ''\n",
        "  for i in range(1, len(RNA_seq_rev)):\n",
        "    cDNAf2 += base_pairing[RNA_seq_rev[i]]\n",
        "  for i in range(0, len(cDNAf2), 3): \n",
        "    codon5 = cDNAf2[i:i + 3]\n",
        "    if len(codon5)==3:\n",
        "      codons5.append(codon5)    \n",
        "  cDNAf3 = ''\n",
        "  for i in range(2, len(RNA_seq_rev)):\n",
        "    cDNAf3 += base_pairing[RNA_seq_rev[i]]\n",
        "  for i in range(0, len(cDNAf3), 3): \n",
        "    codon6 = cDNAf3[i:i + 3]\n",
        "    if len(codon6)==3:\n",
        "      codons6.append(codon6) \n",
        "  return codons1 , codons2 , codons3, codons4, codons5, codons6\n",
        "\n",
        "\n",
        "\n",
        "# I decide to use Dr. Jie Hou to make it easier to code \n",
        "\n",
        "def get_protein_from_DNA(DNA):\n",
        "  mRNA=transcribe(DNA)\n",
        "  mRNA1= mRNA[0]\n",
        "  mRNA2= mRNA[1]\n",
        "  mRNA3= mRNA[2]\n",
        "  mRNA4= mRNA[3]\n",
        "  mRNA5= mRNA[4]\n",
        "  mRNA6= mRNA[5]\n",
        "  one_letter1=''\n",
        "  one_letter2=''\n",
        "  one_letter3=''\n",
        "  one_letter4=''\n",
        "  one_letter5=''\n",
        "  one_letter6=''\n",
        "  \n",
        "  found_start = False\n",
        "  for codon in mRNA1:\n",
        "    if codon =='AUG':   \n",
        "      one_letter1 +=  '<'\n",
        "      found_start = True   # I marked the start codon location so that I can highlight the regions later, same as below\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      if found_start:\n",
        "        one_letter1 +=  '>'\n",
        "        found_start = False   # I marked the stop codon location so that Ican highlight the regions later, same as below\n",
        "      continue\n",
        "    one_letter1 +=  AA_mapping[translate[codon]]\n",
        "  \n",
        "  found_start = False\n",
        "  for codon in mRNA2:\n",
        "    if codon =='AUG':\n",
        "      one_letter2 +=  '<'\n",
        "      found_start = True\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      if found_start:\n",
        "        one_letter2 +=  '>'\n",
        "        found_start = False\n",
        "      continue\n",
        "    one_letter2 +=  AA_mapping[translate[codon]]   \n",
        "    \n",
        "  found_start = False\n",
        "  for codon in mRNA3:\n",
        "    if codon =='AUG':\n",
        "      one_letter3 +=  '<'\n",
        "      found_start = True\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      if found_start:\n",
        "        one_letter3 +=  '>'\n",
        "        found_start = False\n",
        "      continue\n",
        "    one_letter3 +=  AA_mapping[translate[codon]]\n",
        "  \n",
        "  found_start = False\n",
        "  for codon in mRNA4:\n",
        "    if codon =='AUG':\n",
        "      one_letter4 +=  '<'\n",
        "      found_start = True\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      if found_start:\n",
        "        one_letter4 +=  '>'\n",
        "        found_start = False\n",
        "      continue\n",
        "    one_letter4 +=  AA_mapping[translate[codon]]\n",
        "  \n",
        "  found_start = False\n",
        "  for codon in mRNA5:\n",
        "    if codon =='AUG':\n",
        "      one_letter5 +=  '<'\n",
        "      found_start = True\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      if found_start:\n",
        "        one_letter5 +=  '>'\n",
        "        found_start = False\n",
        "      continue\n",
        "    one_letter5 +=  AA_mapping[translate[codon]]\n",
        "  \n",
        "  found_start = False\n",
        "  for codon in mRNA6:\n",
        "    if codon =='AUG':\n",
        "      one_letter6 +=  '<'\n",
        "      found_start = True\n",
        "    if codon =='UAA' or codon=='UGA'or codon =='UAG':\n",
        "      if found_start:\n",
        "        one_letter6 +=  '>'\n",
        "        found_start = False\n",
        "      continue\n",
        "    one_letter6 +=  AA_mapping[translate[codon]]\n",
        "  \n",
        "  if one_letter1.find('M') != -1:\n",
        "    one_letter1 = '*' + one_letter1 \n",
        "  \n",
        "  if one_letter2.find('M') != -1:\n",
        "    one_letter2 = '*' + one_letter2 \n",
        "  \n",
        "  if one_letter3.find('M') != -1:\n",
        "    one_letter3 = '*' + one_letter3 \n",
        "  \n",
        "  if one_letter4.find('M') != -1:\n",
        "    one_letter4 = '*' + one_letter4 \n",
        "  \n",
        "  if one_letter5.find('M') != -1:\n",
        "    one_letter5 = '*' + one_letter5 \n",
        "  \n",
        "  if one_letter6.find('M') != -1:\n",
        "    one_letter6 = '*' + one_letter6 \n",
        "\n",
        "  \n",
        "  r = \" * means ORF sequence; \\n5'to 3' Frame 1:\"+ one_letter1+\"\\n5'to 3' Frame 2:\" + one_letter2 + \"\\n5'to 3' Frame 3:\" + one_letter3 + \"\\n3'to 5' Frame 1:\" + one_letter4 + \"\\n3'to 5' Frame 2:\" + one_letter5 + \"\\n3'to 5' Frame 3:\"+one_letter6 \n",
        "  ## better to concatenate them as one string \n",
        "  return r\n",
        "\n",
        "base_pairing = {'A':'U', 'U':'A', 'C':'G','G':'C'}\n",
        "\n",
        "\n",
        "results = get_protein_from_DNA(test_seq)\n",
        "print (results) \n"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "*\u001b[1m\u001b[94mMSTHDTSLK\n",
            "CLLTIHPK\n",
            "VYSRYIPE\n",
            "*FSG\u001b[1m\u001b[94mMYRE\u001b[0mT\n",
            "FQGCIVSRH\n",
            "FRDVSVD\n",
            " * means ORF sequence; \n",
            "5'to 3' Frame 1:*<MSTHDTSLK\n",
            "5'to 3' Frame 2:CLLTIHPK\n",
            "5'to 3' Frame 3:VYSRYIPE\n",
            "3'to 5' Frame 1:*FSG<MYRE>T\n",
            "3'to 5' Frame 2:FQGCIVSRH\n",
            "3'to 5' Frame 3:FRDVSVD\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nylhNPyQoz0s",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 272
        },
        "outputId": "04bc9b0a-ff01-47ca-9cb0-f7ce1902b78d"
      },
      "source": [
        "\n",
        "'''\n",
        "Do not change codes below.\n",
        "The following codes are used to bring your codes to webpage so that you are able to submit your sequence through website to perform DNA translation task. \n",
        "'''\n",
        "\n",
        "##### web application packages\n",
        "!pip install flask-ngrok\n",
        "from flask_ngrok import run_with_ngrok\n",
        "from flask import Flask, request, render_template\n",
        "! mkdir templates"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting flask-ngrok\n",
            "  Downloading https://files.pythonhosted.org/packages/af/6c/f54cb686ad1129e27d125d182f90f52b32f284e6c8df58c1bae54fa1adbc/flask_ngrok-0.0.25-py3-none-any.whl\n",
            "Requirement already satisfied: Flask>=0.8 in /usr/local/lib/python3.6/dist-packages (from flask-ngrok) (1.1.2)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.6/dist-packages (from flask-ngrok) (2.23.0)\n",
            "Requirement already satisfied: click>=5.1 in /usr/local/lib/python3.6/dist-packages (from Flask>=0.8->flask-ngrok) (7.1.2)\n",
            "Requirement already satisfied: Werkzeug>=0.15 in /usr/local/lib/python3.6/dist-packages (from Flask>=0.8->flask-ngrok) (1.0.1)\n",
            "Requirement already satisfied: Jinja2>=2.10.1 in /usr/local/lib/python3.6/dist-packages (from Flask>=0.8->flask-ngrok) (2.11.2)\n",
            "Requirement already satisfied: itsdangerous>=0.24 in /usr/local/lib/python3.6/dist-packages (from Flask>=0.8->flask-ngrok) (1.1.0)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests->flask-ngrok) (2.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests->flask-ngrok) (2020.6.20)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/dist-packages (from requests->flask-ngrok) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests->flask-ngrok) (3.0.4)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.6/dist-packages (from Jinja2>=2.10.1->Flask>=0.8->flask-ngrok) (1.1.1)\n",
            "Installing collected packages: flask-ngrok\n",
            "Successfully installed flask-ngrok-0.0.25\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xRt1ow2mo8FR",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        },
        "outputId": "50eab72d-c1f2-441d-b75f-db0eb25172e2"
      },
      "source": [
        "######  Write webpage for front-end visualization\n",
        "%%writefile templates/index.html\n",
        "<!doctype html>\n",
        "<html>\n",
        "    <head><link rel=\"stylesheet\" href=\"https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css\"></head>\n",
        "    <body>\n",
        "        <div class=\"container\">\n",
        "            <h1 style=\"text-align:center;\">CSCI1020: Lab05 ----- Build in-class DNA Translation Tool</h1>\n",
        "\n",
        "            <div  style=\"background-color:powderblue;font-size:20px;text-align:center;width:1100px\">\n",
        "            This is a tool which allows the translation of a nucleotide (DNA) sequence to a protein sequence.\n",
        "            <br>\n",
        "            Reference: <a href=\"https://web.expasy.org/translate/\" target=\"_blank\">Official Online tool</a>\n",
        "            </div>\n",
        "\n",
        "            <h2 style=\"text-align:center;\">Upload your DNA sequence:</h2>\n",
        "            \n",
        "            <form method=\"POST\" enctype=\"multipart/form-data\" style=\"margin: 0 auto; text-align: center;\">\n",
        "                <p><textarea name=\"dna_seq\" cols=\"150\" rows=\"10\"></textarea></p>\n",
        "                <input type=\"submit\" value=\"Translate\" style = \"font-size:30px\">\n",
        "            </form>\n",
        "            \n",
        "            <hr>\n",
        "            {% if dna_seq %}\n",
        "            <h2 style=\"text-align:center;\">Submitted DNA Sequence:</h2>\n",
        "\n",
        "            <p><textarea name=\"dna_seq\" cols=\"150\" rows=\"10\" readonly style=\"display: block;margin-left: auto;margin-right: auto;\">{{ dna_seq }}</textarea></p>\n",
        "            \n",
        "            <h2 style=\"text-align:center;\">Results of Translation:</h2>\n",
        "            <p><textarea cols=\"150\" rows=\"5\" readonly style=\"display: block;margin-left: auto;margin-right: auto;\">{{ protein_seq }}</textarea></p>\n",
        "            {% endif %}\n",
        "\n",
        "        </div>\n",
        "    </body>\n",
        "</html>\n"
      ],
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Writing templates/index.html\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KkVTujDBo-yc",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 326
        },
        "outputId": "73bc03ef-98be-4969-8dcd-54fd5df87df4"
      },
      "source": [
        "#####  Implement back-end and front-end application\n",
        "\n",
        "app = Flask(__name__)\n",
        "run_with_ngrok(app)   #starts ngrok when the app is run\n",
        "\n",
        "\n",
        "def home():\n",
        "    return render_template('index.html')\n",
        "\n",
        "@app.route('/', methods=['GET', 'POST'])\n",
        "def index():\n",
        "    if request.method == 'POST':\n",
        "        dna_seq = request.form['dna_seq']\n",
        "        dna_seq = dna_seq.replace('\\n', '').replace('\\r', '').replace(' ', '')\n",
        "        \n",
        "        ''' \n",
        "        Call the function to perform DNA translation\n",
        "        '''\n",
        "        protein_seq = get_protein_from_DNA(dna_seq)\n",
        "\n",
        "        return render_template('index.html',\n",
        "                               dna_seq=dna_seq,\n",
        "                               protein_seq=protein_seq)\n",
        "    else:\n",
        "        return render_template('index.html')\n",
        "\n",
        "app.run() "
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            " * Serving Flask app \"__main__\" (lazy loading)\n",
            " * Environment: production\n",
            "\u001b[31m   WARNING: This is a development server. Do not use it in a production deployment.\u001b[0m\n",
            "\u001b[2m   Use a production WSGI server instead.\u001b[0m\n",
            " * Debug mode: off\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            " * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            " * Running on http://1b559090ee7d.ngrok.io\n",
            " * Traffic stats available on http://127.0.0.1:4040\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "127.0.0.1 - - [15/Sep/2020 02:37:53] \"\u001b[37mGET / HTTP/1.1\u001b[0m\" 200 -\n",
            "127.0.0.1 - - [15/Sep/2020 02:37:55] \"\u001b[33mGET /favicon.ico HTTP/1.1\u001b[0m\" 404 -\n",
            "127.0.0.1 - - [15/Sep/2020 02:38:05] \"\u001b[37mPOST / HTTP/1.1\u001b[0m\" 200 -\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "*GQKKPSPCLLTIHPKPLRKWLFRSSCFASLGLGLLP\u001b[1m\u001b[94mMYFSLSIISLQSRLVLNRGPDK\u001b[0mFDTWLWP\u001b[1m\u001b[94mMP\u001b[0mLSSSLYFQTTLLLQLFLKLTSNVNNSSLASWQEAQTCVQLVFVSISLSHLFLLIQVKEYSEQVSQTWQVILVTVVGSSVSITSTFQLRSLVHSQTITITLKASCSVPLLISVALSSGLP\u001b[1m\u001b[94mM\u001b[1m\u001b[94mMPHS\u001b[0mASWSGPVSPWYFSSIDIVRECSTYSLSIRTPGAKQRPQQPILSCWSHLLAFIFVLFVSSFTPILYILIIPGIA\u001b[1m\u001b[94mMTFWFRVSLQFLLYC\u001b[0mPSETLRVLVLCSSTVESQSHKCQTQKTALLIPLNTLFHKYVFKSLYEQG\u001b[1m\u001b[94mMVLTAILIKE\u001b[0mGYNHLLIKDFWLESDNSELFTTLHS\n",
            "*VRKSPLHVYSRYIPENHGSGFSDHLALPVWGWDFCQCISLCLFLSNLDWFTEAQTSDFKTHGCGQCLNSLPHYISKQHDDFCSNYSSNPQ\u001b[1m\u001b[94mM\u001b[0mIRILHSPRGKKHKLVFNLCSEYPSVCHTCSCFRRNTQSKCHKHGKLFLLQLLVLQCLKHLHSNGHWSTVNRQQLKQVVLFHFFQCRHCLLEVCPCHIHEHHGLDQCLHGTSPPTLSENAVHIHSQSGPQGPSRDHSNPYYPDAGSHICWLLSSKSYLYHLLHLFYIFSSFPEALQHFGFGFPYNFSFTVDLQRPGSLFCVLQLLKARVTKNAKHRRQLCYHILYSIN\u001b[1m\u001b[94mMFLKAC\u001b[1m\u001b[94mMNKVWCSLLYL\u001b[0mKSKVIITCYEKISGWNLIETVSYSPPSIL\n",
            "*SEKALS\u001b[1m\u001b[94mMSTHDTSLKTTEEVAFQIILLCQFGVGTFANVFLFVYNFSPISTGSKQRPRQVILRH\u001b[1m\u001b[94mMAVANALTLFLTIFPNN\u001b[1m\u001b[94mM\u001b[1m\u001b[94mMTFAPIIPQTDLKCKLEFFTRLVARSTNLCSTCVLSIHQFVTLVPVNSGKGILRASVTN\u001b[1m\u001b[94mMASYSCYSCWFFSVLNNIYIPIKVTGPQLTDNNNNSKSKLFCSTSDFSVGIVFLRFAHDATF\u001b[1m\u001b[94mMSI\u001b[1m\u001b[94mMVWTSVS\u001b[1m\u001b[94mMVLLLHRHCQR\u001b[1m\u001b[94mMQYIFTLNQDPRGQAETTATHTIL\u001b[1m\u001b[94mMLVVTFVGFYLLSLICIIFYTYFIYSHHSLRHCNDILVSGFPTISPLLLTFRDPKGPCSVFFNC\u001b[0mKPESLK\u001b[1m\u001b[94mMPNTEDSFANTIKYFIP\u001b[0mICFKLVTRYGAHCYTYKRVRLSLVD\u001b[1m\u001b[94mMKRFLVGI\u001b[0mLKQVIHHPPFS\n",
            "*REWRVVNNSLFQSDSNQKSFHINKLPYSFISIAVSTIPCSYKLLKTYLWNKVFNGISKAVFCVWHFLWLSTVEEHRTRTLRVSEGQQRRNCRETRNQNVIA\u001b[1m\u001b[94mMPQG\u001b[1m\u001b[94mM\u001b[1m\u001b[94mMRIYKIGVKDDTNKT\u001b[0mKIKANKCDYQHQDS\u001b[1m\u001b[94mMGCCGLCLAPGVLIESEYVLHSLT\u001b[1m\u001b[94mMS\u001b[1m\u001b[94mMEEKYHGDTGPDHDAHECGI\u001b[1m\u001b[94mMGKPQEDNAYTEIRSGTEQLAFRVIVIVC\u001b[0mLWTSDLNWNVDVIDTEEPTTVTRITCHVCDTCSEYSFTINRNKCDKL\u001b[1m\u001b[94mMDTQNTS\u001b[0mTQVCASCHEASEEFFTFEVSLRNNWSKSHHVVWKYSEEESGIGHSHVSNHLSGPLFRTSRDWREIIDKEKYIGKSPNPKLAKQDDLKSHFLSGFQGCIVSRHGEGFF\n",
            "*ENGGWITHCFNQIPTRNLFISTSDYNLTLLVQAPYLVHTSFKHIYGIKYL\u001b[1m\u001b[94mMVLAKLSSVFGIFSDSGFQQLKNTEQGPLGSLKVNSKGEIVGKPETK\u001b[1m\u001b[94mMSLQCLRE\u001b[0mEYIKVK\u001b[1m\u001b[94mMIQIRLRR\u001b[0mKPTNVTTSIRIVWVAVVSAWPLGSLRVN\u001b[1m\u001b[94mMYCIL\u001b[0mQCLWRRST\u001b[1m\u001b[94mMETLVQT\u001b[1m\u001b[94mM\u001b[1m\u001b[94mML\u001b[1m\u001b[94mMNVASWANLKKT\u001b[1m\u001b[94mMPTLKSEVEQNNLLLELLLLSVNCGPVTLIG\u001b[1m\u001b[94mM\u001b[0m\u001b[1m\u001b[94mMLFKTLKNQQL\u001b[0mQELA\u001b[1m\u001b[94mMFVTLALSIPLPELTGTSVTN\u001b[0mWILRTQVEHKFVLLATRRVKNSNLHLRSVGIIGAKVI\u001b[1m\u001b[94mMLFGNIVRKRVKALATA\u001b[1m\u001b[94mMCLKITCLGLCLEPVEIGEKL\u001b[0mTKRNTLAKVPTPNWQSK\u001b[1m\u001b[94mMI\u001b[0mKATSSVVFRDVSVD\u001b[1m\u001b[94mMERAFSD\n",
            "*R\u001b[1m\u001b[94mMEGGE\u001b[0mLTVSIRFQPEIFSYQQVIITLLFYKYSSEHHTLFIQAFKNIF\u001b[1m\u001b[94mME\u001b[0mSIWYQSCLLCLAFLVTLAFNSRTQNKDPGLRSTVKEKLGNPKPKCHCNASGNDENINRCKRYKDLEDKSQQ\u001b[1m\u001b[94mM\u001b[0mLPASGYGLLWSLLGPWGPDEICTAFSDNVYGGEVPWRHWSRPCS\u001b[1m\u001b[94mMWHHGQTSRRQCLH\u001b[0mNQKWNRTTCFSYCYCLLTVDQPLECRCYLRHRTNNCNKNNLPCLHLLVFLYLNQEQVQTDGYSEHKLNTSLCFLPRGERILIYIGQFEELEQKSSCCLEIGRELRHWPQPCVLKSLVWASVNQSRLERNYRQREIHWQKSQPQTGKARSEKPLPQWFSG\u001b[1m\u001b[94mMYRE\u001b[0mTWRGLFLT\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}